vllm/utils.py

```diff
@@ -922,3 +922,215 @@ def parse_args(self, args=None, namespace=None):
+            "I8": torch.int8,
+            "U8": torch.uint8,
+            "I16": torch.int16,
+            "U16": torch.uint16,
+            "I32": torch.int32,
+            "U32": torch.uint32,
+            "I64": torch.int64,
+            "U64": torch.uint64,
+            "F16": torch.float16,
+            "F32": torch.float32,
+            "F64": torch.float64,
+            "BF16": torch.bfloat16,
+            "F8_E4M3": torch.float8_e4m3fn,
+            "F8_E5M2": torch.float8_e5m2
+        }
+        dtype = type_mapping[slice_view.get_dtype()]
+        shape = tuple(slice_view.get_shape())
+        if shape:
+            self._meta_tensor = torch.zeros(*shape, dtype=dtype, device="meta")
+        else:
+            self._meta_tensor = torch.zeros(tuple(),
+                                            dtype=dtype,
+                                            device="meta")
+
+    def __getattr__(self, name):
+        if name in ["shape", "dtype", "size", "stride"]:
+            # redirect metadata information queries to the meta tensor
+            return getattr(self._meta_tensor, name)
+        if hasattr(torch.Tensor, name):
+            # the rest functions will materialize the tensor and call the
+            # function on the materialized tensor
+            tensor = self.materialize()
+            return getattr(tensor, name)
+        raise AttributeError(f"Attribute {name} not found")
+
+    def __getitem__(self, key) -> torch.Tensor:
+        return self._slice_view[key]
+
+    def materialize(self) -> torch.Tensor:
+        return self._slice_view[tuple()]
+
+    def narrow(input, dim, start, length) -> torch.Tensor:
+        # `input` is a `DeferredTensor` object
+        # it does not use `self`, but `input` instead
+        # to better match https://pytorch.org/docs/stable/generated/torch.narrow.html signature # noqa
+
+        # `DeferredTensor` will only respond to `narrow` method
+        # which reads the corresponding data from disk and returns
+        # a materialized tensor
+        slices = [slice(None, None, None) for x in input._meta_tensor.shape]
+        slices[dim] = slice(start, start + length)
+        return input._slice_view[tuple(slices)]
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func == torch.narrow:
+            if len(args) >= 2:
+                kwargs["dim"] = args[1]
+            if len(args) >= 3:
+                kwargs["start"] = args[2]
+            if len(args) >= 4:
+                kwargs["length"] = args[3]
+            return args[0].narrow(**kwargs)
+        new_args = []
+        for arg in args:
+            if isinstance(arg, DeferredTensor):
+                new_args.append(arg.materialize())
+            else:
+                new_args.append(arg)
+        return func(*new_args, **kwargs)
+
+    # implement common tensor operations, except for in-place operations
```

Comment:
why do we need those in the first place?

Reply:
I write these code so that the deferred tensors can behave just like normal tensors. for example, some users call `torch.nn.functional.normalize` on tensors loaded from disk : 

https://github.com/vllm-project/vllm/blob/543aa4857362ff385af8a9f496392b8831c42833/vllm/model_executor/models/baichuan.py#L382

When we have these functions, users' code will be minimally affected.

Reply:
I think this will lead to issues later down the line (we now have an ill-defined subset of supported operations). I feel like explicitly requiring materialization is a better way.

Reply:
you can refer to [da9d486](https://github.com/vllm-project/vllm/pull/6127/commits/da9d486cdc4a8d58133b1a7f0aec2332436f67fe) for example. that is what i did before. I find that code is more intrusive, and all third-party code using vllm might break.

Reply:
I think this is too magical and being explicit would be better, but I am not going to block on this.

Reply:
@Yard1 would you consider elaborating. I am not sure I agree that explicitly presenting the tensor `bytearrray` content in a person's mental model ordered wrt consumption of said `bytearray` is wise, especially for sarge a large read-only buffer. I am of opinion that any large read-only memory is a massive target for performance optimization.

I have been working on a very small PR designed to improve availability. This is done by eliminating the model load startup delay (about 1 minute on my gear, I can shave it to about 3 seconds). This is pretty straightforward to deliver, presuming that the logit operation does not need to be calculated on startup.

The gist of my WIP PR is:
- remove logits calculation (which takes 12 seconds on my very beast GPU server with 2 A40 running llama3-70b-w8a8.
- Rely on safetensors exclusively.
- Modify, slightly, the management of pytorch `storage` such that IF a `storage` buffer address is accessed with any Tensor.to('cuda'), there will be some intelligence as it relates to h2d memcpy..
- Modify, slightly, to('cuda') calls to send Tensor objects to cuda devices, but not to execute a cuda host-to-device copy (as is default with Pytorch).

Would be interested in hearing your opinions on the approach and a review when ready. There are some constraints:
- safetensors first implementation
- cuda first implementation
- cuda must have `uvm` loaded
- dedicated address space consumption of model size
- dedicated (initially) memory consumption of model size (this is an obnoxious constraint, and is not permanent).
- no run-time quantization. (I prefer static compressed models, although I understand the value of some types of dynamic calculations of decompression, so we can kick this around).

One problem I struggle with is not technical - its about making this model an optional feature (via runtime environment variable), or adding as a flag.

Thanks that was helpful writing down my ideas in this area, I will copy this into an issue tracker. 
cc/ @sdake
cc/ @robertgshaw2-neuralmagic